python/hail/linalg/blockmatrix.py

```diff
@@ -553,103 +584,281 @@ def __neg__(self):
         op = getattr(self._jbm, "unary_$minus")
         return BlockMatrix(op())
 
-    @typecheck_method(b=oneof(numeric, block_matrix_type))
+    # scalar => 0, column => 1, row => 2, else => 3
+    @staticmethod
+    def _shape_type(shape):
+        assert len(shape) == 2
+        if shape[0] == 1 and shape[1] == 1:
+            return 0
+        elif shape[1] == 1:
+            return 1
+        elif shape[0] == 1:
+            return 2
+        else:
+            return 3
+
+    @staticmethod
+    def _check_shapes(shape, shape2, op='block matrix operation'):
+        st = BlockMatrix._shape_type(shape)
+        st2 = BlockMatrix._shape_type(shape2)
+
+        if (st == 0 or st2 == 0 or (st == st2 and shape == shape2) or
+                (((st == 3 and st2 == 1) or (st == 1 and st2 == 3)) and shape[0] == shape2[0]) or
+                (((st == 3 and st2 == 2) or (st == 2 and st2 == 3)) and shape[1] == shape2[1])):
+            return st, st2
+        else:
+            raise ValueError(f'incompatible shapes for {op}: {shape} and {shape2}')
+
+    @typecheck_method(b=oneof(numeric, np.ndarray, block_matrix_type))
     def __add__(self, b):
         """Addition: a + b.
 
         Parameters
         ----------
-        b: :class:`BlockMatrix` or :obj:`int` or :obj:`float`
+        b: :obj:`int` or :obj:`float` or :class:`numpy.ndarray` or :class:`BlockMatrix`
 
         Returns
         -------
         :class:`.BlockMatrix`
         """
+        op = 'addition'
         if isinstance(b, int) or isinstance(b, float):
             return BlockMatrix(self._jbm.scalarAdd(float(b)))
+        elif isinstance(b, np.ndarray):
```

Comment:
I'm not 100% sure how to write this but I think we can do better.  An initial suggestion:

have a `new_self, new_b = _promote(self, b)` such that self and new_b are "valid" to be combined.  You might need a few of these depending on the operation (e.g. add vs sub vs matmul).  Maybe it also returns a flag whether the operands were reversed.

This should lead to 4 cases: new_self is always a BlockMatrix, new_b is a scalar, (n, 1) or (1, m) ndarray or an (n, m) block matrix (or 8, and they might be reversed).